### ISSUE TITLE
Speed Regression in torch.qr

### ISSUE THREAD

**Issue by jacobrgardner** [pytorch/pytorch#23580]
## 🐛 Bug

I'm seeing some pretty serious (~5-10x) speed regressions for `torch.qr` on master compared to the anaconda release with the following test code:
```python
a = torch.randn(1000, 10)
%timeit q, r = torch.qr(a)
```
On release PyTorch:
```
253 ms ± 274 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
And on master:
```                                                                                                                                                                                 
2.66 s ± 509 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by SsnL** [comment]
cc @vishwakftw could this be related to your QR batching PR? Was that in 1.1?

**Comment by vishwakftw** [comment]
QR batching PR was done after 1.1, please allow me to investigate.

**Comment by vishwakftw** [comment]
@jacobrgardner These timings look really skewed to me. I benchmarked torch.qr on two versions of PyTorch - `v1.1` and `v1.2.0.dev20190730+cpu`. The consolidated timings are given [here](https://gist.github.com/vishwakftw/2ee0073e8273a97ad925ff75eefc9c9d.

TL;DR: there is a minor regression (about 1.01 - 1.1). I'll try to fix this as soon as I can.

**Comment by jacobrgardner** [comment]
@vishwakftw Hmm, timing differences that small wouldn't have raised alarm bells on my end -- the symptom that led me to discovering this was an actual model training around 10x slower than it ought to have with `torch.qr` dominating the running time.

I'll look in to this on my end in the morning and see if it was something local, but the problem was consistent enough that I ended up hotfixing out my usage `torch.qr` in favor of the Woodbury formula for the moment.

**Comment by vishwakftw** [comment]
@jacobrgardner sure, thank you. I have a local fix for the regression, will run tests and create a PR accordingly.

**Comment by vishwakftw** [comment]
[`line_profiler`](https://pypi.python.org/pypi/line_profiler) outputs the below profile. It seems like the `pivoted_cholesky` call seems to take the most time.

```
Timer unit: 1e-06 s

Total time: 0.021249 s
File: /media/vishwak/Official-1/gpytorch/gpytorch/lazy/added_diag_lazy_tensor.py
Function: _preconditioner at line 53

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    53                                               def _preconditioner(self):
    54         7         53.0      7.6      0.2          if settings.max_preconditioner_size.value() == 0:
    55                                                       return None, None, None
    56                                           
    57         7         31.0      4.4      0.1          if not (hasattr(self, "_woodbury_cache") or hasattr(self, "self._q_cache")):
    58         7         24.0      3.4      0.1              max_iter = settings.max_preconditioner_size.value()
    59         7      17972.0   2567.4     84.6              self._piv_chol_self = pivoted_cholesky.pivoted_cholesky(self._lazy_tensor, max_iter)
    60         7        275.0     39.3      1.3              if torch.any(torch.isnan(self._piv_chol_self)).item():
    61                                                           warnings.warn(
    62                                                               "NaNs encountered in preconditioner computation. Attempting to continue without preconditioning."
    63                                                           )
    64                                                           return None, None, None
    65                                           
    66                                                       # TODO: Hotfix removing use of QR due to drastic speed regression in PyTorch (note that I removed the False)
    67         7         21.0      3.0      0.1              if self._piv_chol_self.dim() == 2:  # TODO: Whenever PyTorch supports batch mode
    68         7         29.0      4.1      0.1                  *batch_shape, n, k = self._piv_chol_self.shape
    69         7         94.0     13.4      0.4                  self._noise = self._diag_tensor.diag().unsqueeze(-1)
    70                                           
    71         7        257.0     36.7      1.2                  self.constant_diag = torch.equal(self._noise, self._noise[0] * torch.ones_like(self._noise))
    72                                           
    73         7        122.0     17.4      0.6                  eye = torch.eye(k, dtype=self._piv_chol_self.dtype, device=self._piv_chol_self.device)
    74                                           
    75         7         14.0      2.0      0.1                  if self.constant_diag:
    76                                                               # We can factor out the noise for for both QR and solves.
    77                                                               self.noise_constant = self._noise[0].squeeze()
    78                                                               self._q_cache, self._r_cache = torch.qr(
    79                                                                   torch.cat((self._piv_chol_self, self.noise_constant.sqrt() * eye))
    80                                                               )
    81                                                               self._q_cache = self._q_cache[:n, :]
    82                                           
    83                                                               # Use the matrix determinant lemma for the logdet, using the fact that R'R = L_k'L_k + s*I
    84                                                               logdet = self._r_cache.diagonal(dim1=-1, dim2=-2).abs().log().sum(-1).mul(2)
    85                                                               logdet = logdet + (n - k) * self.noise_constant.log()
    86                                                               self._precond_logdet_cache = logdet.view(*batch_shape) if len(batch_shape) else logdet.squeeze()
    87                                           
    88                                                           else:
    89                                                               # With non-constant diagonals, we cant factor out the noise as easily
    90         7        806.0    115.1      3.8                      self._q_cache, self._r_cache = torch.qr(torch.cat((self._piv_chol_self / self._noise.sqrt(), eye)))
    91         7        225.0     32.1      1.1                      self._q_cache = self._q_cache[:n, :] / self._noise.sqrt()
    92                                           
    93         7        436.0     62.3      2.1                      logdet = self._r_cache.diagonal(dim1=-1, dim2=-2).abs().log().sum(-1).mul(2) - (
    94         7        298.0     42.6      1.4                          1.0 / self._noise
    95         7        158.0     22.6      0.7                      ).log().sum([-1, -2])
    96         7         74.0     10.6      0.3                      self._precond_logdet_cache = logdet.view(*batch_shape) if len(batch_shape) else logdet.squeeze()
    97                                           
    98                                                       else:
    99                                                           self._woodbury_cache, self._inv_scale, self._precond_logdet_cache = woodbury.woodbury_factor(
   100                                                               self._piv_chol_self, self._piv_chol_self, self._diag_tensor.diag(), logdet=True
   101                                                           )
   102                                                           self._scaled_inv_diag = self._inv_scale / self._diag_tensor.diag().unsqueeze(-1)
   103                                                           self._scaled_inv_diag_piv_chol_self = self._piv_chol_self * self._scaled_inv_diag
   104                                           
   105         7        335.0     47.9      1.6              self.preconditioner_lt = PsdSumLazyTensor(RootLazyTensor(self._piv_chol_self), self._diag_tensor)
   106                                           
   107                                                   # NOTE to future self:
   108                                                   # We cannot memoize this precondition closure
   109                                                   # It causes a memory leak otherwise
   110         7         14.0      2.0      0.1          def precondition_closure(tensor):
   111                                                       if hasattr(self, "_q_cache"):
   112                                                           if self.constant_diag:
   113                                                               return (1 / self.noise_constant) * (tensor - self._q_cache.matmul(self._q_cache.t().matmul(tensor)))
   114                                           
   115                                                           else:
   116                                                               return (tensor / self._noise) - self._q_cache.matmul(self._q_cache.t().matmul(tensor))
   117                                           
   118                                                       else:
   119                                                           res = woodbury.woodbury_solve(
   120                                                               tensor,
   121                                                               self._scaled_inv_diag_piv_chol_self,
   122                                                               self._woodbury_cache,
   123                                                               self._scaled_inv_diag,
   124                                                               self._inv_scale,
   125                                                           )
   126                                                           return res
   127                                           
   128         7         11.0      1.6      0.1          return precondition_closure, self.preconditioner_lt, self._precond_logdet_cache
```

**Comment by jacobrgardner** [comment]
Right, `pivoted_cholesky` being the majority of the running time would be the expected behavior for this portion of code. What I was seeing earlier instead was not only `torch.qr` taking a majority of the time in this subsection, but a majority of the entire training process due to taking multiple seconds per run.

I've been able to reproduce the problem locally, and interestingly `torch.qr` seems to be the only major linear algebra op affected moving between the two versions. I'm wondering if there's something strange with my installation of something. I'll dig more deeply in to that tomorrow.

**Comment by vishwakftw** [comment]
Could you give a reproduction script for reference is possible? I’ll also try to diagnose the issue from my end.

**Comment by vishwakftw** [comment]
I’ll keep this issue open for discussion.
